controller/backup_controller.go

```diff
@@ -413,153 +415,122 @@ func (bc *BackupController) validateBackingImageChecksum(volName, biName string)
-		backup.Status.SnapshotName = backupStatus.SnapshotName
-		backup.Status.ReplicaAddress = backupStatus.ReplicaAddress
-		backup.Status.State = backupStatus.State
-		return true, nil
-	case longhorn.BackupStateCompleted:
-		if bm := bc.hasMonitor(backup.Name); bm != nil {
-			backupStatus := bm.GetBackupStatus()
+	// Backing image checksum validation
+	biChecksum, err := bc.validateBackingImageChecksum(volumeName, volume.Spec.BackingImage)
+	if err != nil {
+		return err
+	}
 
-			backup.Status.Progress = backupStatus.Progress
-			backup.Status.URL = backupStatus.URL
-			backup.Status.Error = backupStatus.Error
-			backup.Status.SnapshotName = backupStatus.SnapshotName
-			backup.Status.ReplicaAddress = backupStatus.ReplicaAddress
-			backup.Status.State = backupStatus.State
+	// Initialize a backup target client
+	backupTargetClient, err := getBackupTargetClient(bc.ds, backupTarget)
+	if err != nil {
+		return err
+	}
 
-			// Trigger updates backup volume to run reconcile immediately
-			syncTime := metav1.Time{Time: time.Now().UTC()}
-			backupVolume, err := bc.ds.GetBackupVolume(volume.Name)
-			if err == nil {
-				// Request backup_volume_controller to reconcile BackupVolume immediately.
-				backupVolume.Spec.SyncRequestedAt = syncTime
-				if _, err = bc.ds.UpdateBackupVolume(backupVolume); err != nil && !apierrors.IsConflict(errors.Cause(err)) {
-					log.WithError(err).Errorf("Error updating backup volume %s spec", volume.Name)
-				}
-			} else {
-				if !apierrors.IsNotFound(err) {
-					log.WithError(err).Warnf("Cannot get the backup volume %s", volume.Name)
-				}
-
-				// Request backup_target_controller to reconcile BackupTarget immediately.
-				backupTarget.Spec.SyncRequestedAt = syncTime
-				if _, err = bc.ds.UpdateBackupTarget(backupTarget); err != nil && !apierrors.IsConflict(errors.Cause(err)) {
-					log.WithError(err).Warn("Failed to update backup target")
-				}
-			}
-		}
+	// Find the corresponding engine client
+	engineClient, err := bc.getEngineClient(volumeName)
+	if err != nil {
+		return err
+	}
 
-		// Disable backup monitor
-		bc.disableBackupMonitor(backup)
-		return false, nil
-	case longhorn.BackupStateError, longhorn.BackupStateUnknown:
-		// Disable backup monitor
-		bc.disableBackupMonitor(backup)
-		return true, nil
+	// Call engine API snapshot backup
+	_, replicaAddress, err := engineClient.SnapshotBackup(backup.Name, backup.Spec.SnapshotName, backupTargetClient.URL, volume.Spec.BackingImage, biChecksum, backup.Spec.Labels, backupTargetClient.Credential)
+	if err != nil {
+		log.WithError(err).Warn("Cannot take snapshot backup")
+		backup.Status.Error = err.Error()
+		backup.Status.State = longhorn.BackupStateError
+		return err
 	}
+	backup.Status.ReplicaAddress = replicaAddress
+	backup.Status.State = longhorn.BackupStateInProgress
 
-	log.Warnf("Unknown backup state %s", backup.Status.State)
-	return true, nil
+	monitor = bc.enableBackupMonitor(backup, volumeName, engineClient, replicaAddress)
```

Comment:
I am thinking if we can move this API call into the monitor: When the monitor is launched and the backup current state is New, the API will be invoked. And in the next loop monitor could get the state from the engine/replica so that the controller does not need to update the state manually here.

Reply:
I had thought this before. The result would be that if the backup size is small and the backup process is finished within the poll interval (2 seconds here), then the event log will only record the backup transition state `Completed` since the state changed from New -> Completed.
If the backup process time is over than the poll interval (2 seconds), then the event log will have transition state `InProgress` and `Completed`.

If we're don't care the event log must have InProgress -> Completed state, then yes we could move the `engineClient.SnapshotBackup` API call into the `NewBackupMonitor`.

Reply:
I don't think a backup should restrictedly follow New -> InProgress -> Completed flow. What the controller does here is **blindly** syncing the state from the engine/replica. This `blind` means we cannot control the flow.

Reply:
Updated.